File: scala/org/softlang/company/Subunit.scala
package org.softlang.company

/**
 * A subunit is either a person unit or a sub department.
 */
class Subunit {
  private var pu: Employee = null
  private var du: Dept = null

  def this(pu: Employee) = {
    this()
    this.pu = pu
  }
  def this(du: Dept) = {
    this()
    this.du = du
  }

  def getPu = pu
  def setPu(pu: Employee) = {
    this.pu = pu
    this.du = null
  }

  def getDu = du
  def setDu(du: Dept) = {
    this.pu = null
    this.du = du
  }

  def total: Double = if (getPu != null) getPu.total else if (getDu != null) getDu.total else 0.0

  def cut: Unit = {
    if (getPu != null)
      getPu.cut
    if (getDu != null)
      getDu.cut

  }
}

File: java/org/softlang/company/Subunit.java
package org.softlang.company;

import java.io.Serializable;

/**
 * A subunit is either an employee unit or a department unit
 * 
 */
public class Subunit implements Serializable {

	private static final long serialVersionUID = -2008895922137165250L;

	private Employee eUnit;
	private Department dUnit;

	public Employee getEu() {
		return eUnit;
	}

	public void setEu(Employee eu) {
		dUnit = null;
		this.eUnit = eu;
	}

	public Department getDu() {
		return dUnit;
	}

	public void setDu(Department du) {
		eUnit = null;
		this.dUnit = du;
	}

	public double total() {
		double total = 0;
		total += getDu() == null ? 0 : getDu().total();
		total += getEu() == null ? 0 : getEu().total();
		return total;	
	}	
	
	public void cut() {
		if (getDu() != null) getDu().cut();
		if (getEu() != null) getEu().cut();
	}	
}
